scripts/validate_config.py

```diff
@@ -0,0 +1,136 @@
+        help='Path to a validation schema.',
+        required=True,
+        type=cli_helpers.path_if_file_exists,
+    )
+
+    parser.add_argument(
+        '-c', '--config_file',
+        help='Path to configuration file. Accepts YAML, bash/ini or namelist',
+        required=True,
+        type=cli_helpers.path_if_file_exists,
+    )
+
+    parser.add_argument(
+        '--config_file_type',
+        help='If used, will convert provided config file to given file type.\
+            Accepts YAML, bash/ini or namelist',
+        choices=["YAML", "INI", "F90"],
+    )
+
+    group.add_argument(
+        '-v', '--verbose',
+        action='store_true',
+        help='If provided, print all logging messages.',
+        )
+    group.add_argument(
+        '-q', '--quiet',
+        action='store_true',
+        help='If provided, print no logging messages',
+        )
+    parser.add_argument(
+        '-l', '--log_file',
+        help='Optional path to a specified log file',
+        default=os.path.join(os.path.dirname(__file__), "set_config.log")
+        )
+
+    return parser.parse_args(argv)
+
+
+def validate_config(argv, log=None):
+    '''Main section for validating config file'''
+
+    user_args = parse_args(argv)
+
+    if log is None:
+        name = f"{inspect.stack()[0][3]}"
+        log = cli_helpers.setup_logging(user_args, log_name=name)
+
+    # Get the config file to be validated and dereference jinja templates
+    config_class = getattr(config, "YAMLConfig")
+    config_obj = config_class(user_args.config_file, log_name=log.name)
+    config_obj.dereference_all()
+
+    # Load the json validation schema
+    with open(user_args.validation_schema, 'r',
+              encoding="utf-8") as schema_file:
+        schema = json.load(schema_file)
+
+    schema_error = 0
+
+    # Validate the config file against the schema file
+    validator = jsonschema.Draft7Validator(schema)
+
+    # Print out each schema error
+    errors = validator.iter_errors(config_obj.data)
+
+    for error in errors:
+        schema_error += 1
+        logging.error(error)
+        logging.error('------')
+
+    # Create a list of fields that could contain a file or path
+    path_list = []
+    for field in schema["properties"]:
+        for value in schema["properties"][field]["properties"]:
+            if "format" in schema["properties"][field]["properties"][value]:
+                path_list.append(value)
+
+    # Check for existence of those files or paths
+    for field in config_obj.data:
+        for value in config_obj.data[field]:
+            if value in path_list:
+                if not os.path.exists(config_obj.data[field][value]):
+                    schema_error += 1
+                    logging.error('%s has Invalid Path %s',
+                                  value, config_obj.data[field][value])
+
+    if schema_error > 0:
+        sys.exit(f'This configuration file has {str(schema_error)} errors')
+    else:
+        sys.exit(0)
```

Comment:
I think the `str()` wrapper is unnecessary (f-strings will do the `int` -> `str` conversion).

I think it would be better if this function returned a value and left the exiting to the caller, as its stated responsibility is to validate a config, not to terminate the program. IMO it should only validate the config and return a `bool` indicating whether or not it was valid. It should not set up logging: The caller -- or some ancestor -- should do that.

It might be useful to even split this function into two: One that validates the config vs the schema, and one that validates the path values. I can imagine that other correctness checks might be needed later, and the obvious path given the current code will be to add more complexity to this function. Replacing this function with two smaller, single-responsibility functions would suggest a cleaner path for maintainers / extenders to follow.

But I know that @venitahagerty needs to be done with this, so this is probably a suggestion for later work.

Reply:
The python program validate_config.py is a command line script that will be part of a workflow.  It needs to provide a return code that can be tested to see whether the workflow should terminate or continue.

I'll remove the `str()` wrapper.

Reply:
I should probably add that I do realize that this code is following existing idioms, so my comments are really about the idioms, not about this specific instance.

Reply:
Understood, but how about e.g.
```
try:
    if not validate_config(sys.argv[1:], cli_log):
        sys.exit(1)
...
```
? Then CLI interface -- getting/parsing command-line arguments, setting exit status -- lives in the outermost scope. I'm not suggesting that the _script_ shouldn't be terminating the program (it must!) but that a function responsible for validating a config shouldn't.

Reply:
I see. Exit not in validate_config(), but in main. I think your idea of returning a boolean and testing for it in main is good.

Reply:
Just wanted to make sure I wasn't missing anything for when I bring it up again with the people who are still working on this codebase after this week :)